arangod/RocksDBEngine/RocksDBVPackIndex.cpp

```diff
@@ -307,143 +463,402 @@ class RocksDBVPackIndexIterator final : public IndexIterator {
-
-      if constexpr (unique) {
-        cb(RocksDBValue::documentId(_iterator->value()));
-      } else {
-        cb(RocksDBKey::indexDocumentId(_iterator->key()));
-      }
+      consumeIteratorValue();
 
       if (!advance()) {
         // validate that Iterator is in a good shape and hasn't failed
         arangodb::rocksutils::checkIteratorStatus(_iterator.get());
         return false;
       }
 
-      --limit;
-      if (limit == 0) {
+      if (--limit == 0) {
         return true;
       }
     } while (true);
   }
 
-  bool nextCoveringImpl(CoveringCallback const& cb, size_t limit) override {
-    ensureIterator();
-    TRI_ASSERT(_trx->state()->isRunning());
-    TRI_ASSERT(_iterator != nullptr);
-
-    if (limit == 0 || !_iterator->Valid() || outOfRange()) {
-      // No limit no data, or we are actually done. The last call should have
-      // returned false
-      TRI_ASSERT(limit > 0);  // Someone called with limit == 0. Api broken
-      // validate that Iterator is in a good shape and hasn't failed
-      arangodb::rocksutils::checkIteratorStatus(_iterator.get());
-      return false;
-    }
+  void fillResultBuilder() {
+    TRI_ASSERT(_cache != nullptr);
 
-    TRI_ASSERT(limit > 0);
+    _resultBuilder.clear();
+    _resultBuilder.openArray(true);
 
-    do {
+    while (true) {
       rocksdb::Slice key = _iterator->key();
       TRI_ASSERT(_index->objectId() == RocksDBKey::objectId(key));
 
       if constexpr (unique) {
         LocalDocumentId const documentId(
             RocksDBValue::documentId(_iterator->value()));
+        _resultBuilder.add(VPackValue(documentId.id()));
+        _resultBuilder.add(RocksDBKey::indexedVPack(key));
         if (_index->hasStoredValues()) {
-          auto data = SliceCoveringDataWithStoredValues(
-              RocksDBKey::indexedVPack(key),
+          _resultBuilder.add(
               RocksDBValue::uniqueIndexStoredValues(_iterator->value()));
-          cb(documentId, data);
-        } else {
-          auto data = SliceCoveringData(RocksDBKey::indexedVPack(key));
-          cb(documentId, data);
         }
       } else {
         LocalDocumentId const documentId(RocksDBKey::indexDocumentId(key));
+        _resultBuilder.add(VPackValue(documentId.id()));
+        _resultBuilder.add(RocksDBKey::indexedVPack(key));
         if (_index->hasStoredValues()) {
-          auto data = SliceCoveringDataWithStoredValues(
-              RocksDBKey::indexedVPack(key),
+          _resultBuilder.add(
               RocksDBValue::indexStoredValues(_iterator->value()));
-          cb(documentId, data);
-        } else {
-          auto data = SliceCoveringData(RocksDBKey::indexedVPack(key));
-          cb(documentId, data);
         }
       }
 
       if (!advance()) {
         // validate that Iterator is in a good shape and hasn't failed
         arangodb::rocksutils::checkIteratorStatus(_iterator.get());
-        return false;
+        break;
       }
+    }
```

Comment:
What if we have millions of documents with matching index attributes? Do we want to limit the max size of a single cache entry? Otherwise we might build up a huge entry here that will be removed again immediately because we exceed the max cache size.

Reply:
I agree it would be bad to store a huge result in the cache for an index value with many matches.
Right now we do the following:
- look up the value in the cache
- if value not in cache, look up the values in RocksDB
- iterate over the values from RocksDB and store them in `_resultBuilder`
- at the end, store the value from `_resultBuilder` in the cache
We can avoid the last step if the value is too large (whatever size threshold we choose). 
This would have the advantage of not polluting the cache with huge results.
However, whenever such value is looked up, we would need to consult the cache again, only to find that the value is not cached, then fetch it from RocksDB etc... So we will have an extra cache lookup for these values. 
I think that is tolerable given that we are talking about huge values, so the cost of an extra O(1) cache lookup should not matter much.